File: frontend/lib/api/classroom.ts
import { apiClient } from "./client";
import {
  ClassroomNote,
  ClassroomProgress,
  LessonProgressPayload,
  LessonProgressResult,
  UpsertNotePayload,
} from "@/types/classroom";

export const classroomApi = {
  getNote: async (lessonId: string): Promise<ClassroomNote | null> => {
    const { data } = await apiClient.get<ClassroomNote | null>("/notes", {
      params: { lesson_id: lessonId },
    });
    return data;
  },

  upsertNote: async (payload: UpsertNotePayload): Promise<ClassroomNote> => {
    const { data } = await apiClient.post<ClassroomNote>("/notes", payload);
    return data;
  },

  getNotesByCourse: async (courseId: string): Promise<ClassroomNote[]> => {
    const { data } = await apiClient.get<ClassroomNote[]>("/notes/by-course", {
      params: { course_id: courseId },
    });
    return data;
  },

  patchLessonProgress: async (
    lessonId: string,
    payload: LessonProgressPayload,
  ): Promise<LessonProgressResult> => {
    const { data } = await apiClient.patch<LessonProgressResult>(
      `/lessons/${lessonId}/progress`,
      payload,
    );
    return data;
  },

  getClassroomProgress: async (
    courseId: string,
  ): Promise<ClassroomProgress> => {
    const { data } = await apiClient.get<ClassroomProgress>(
      `/classroom/${courseId}/progress`,
    );
    return data;
  },
};

File: backend/app/api/v1/classroom.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.dependencies import get_current_user, get_db
from app.models.user import User
from app.schemas.classroom import (
    ClassroomProgressOut,
    LessonProgressPatchIn,
    LessonProgressPatchOut,
    NoteOut,
    NoteUpsertIn,
)
from app.services.classroom_service import ClassroomService

router = APIRouter(tags=["Classroom"])


@router.get("/notes", response_model=NoteOut | None)
def get_note(
    lesson_id: str = Query(...),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return ClassroomService(db).get_note(current_user, lesson_id)


@router.post("/notes", response_model=NoteOut)
def upsert_note(
    payload: NoteUpsertIn,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return ClassroomService(db).upsert_note(current_user, payload)


@router.get("/notes/by-course", response_model=list[NoteOut])
def get_notes_by_course(
    course_id: str = Query(...),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return ClassroomService(db).get_notes_by_course(current_user, course_id)


@router.patch("/lessons/{lesson_id}/progress", response_model=LessonProgressPatchOut)
def patch_lesson_progress(
    lesson_id: str,
    payload: LessonProgressPatchIn,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return ClassroomService(db).update_lesson_progress(
        current_user,
        lesson_id,
        payload,
    )


@router.get("/classroom/{course_id}/progress", response_model=ClassroomProgressOut)
def get_classroom_progress(
    course_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return ClassroomService(db).get_course_progress(current_user, course_id)
